File: src/commands/scp.ts
import { resolveProjectRoot, ensureClrunDirs } from '../utils/paths';
import { success, fail } from '../utils/output';
import { acquireLock } from '../runtime/lock-manager';
import { generateTerminalId, writeSession } from '../pty/pty-manager';
import { initQueue } from '../queue/queue-engine';
import { initBuffer, appendToBuffer } from '../buffer/buffer-manager';
import { logEvent } from '../ledger/ledger';
import { SCPClient, formatCliForBuffer } from '../scp/client';
import type { SessionMetadata } from '../types';

/**
 * Connect to an SCP server and start a run. Creates a virtual terminal session backed by SCP (no PTY).
 * CLRUN supports dynamic remote CLIs via SCP: run `clrun scp <url>`, then use clrun <id> "<input>" to drive the flow.
 */
export async function scpConnectCommand(baseUrl: string): Promise<void> {
  const projectRoot = resolveProjectRoot();
  ensureClrunDirs(projectRoot);
  acquireLock(projectRoot);

  const terminalId = generateTerminalId();
  initQueue(terminalId, projectRoot);
  initBuffer(terminalId, projectRoot);

  const client = new SCPClient(baseUrl);
  let frame;
  let cli;
  try {
    frame = await client.startRun();
    cli = await client.getCli();
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    fail({
      error: `SCP connect failed: ${msg}`,
      hints: {
        check_url: 'Ensure the SCP server is running and reachable.',
        example: 'clrun scp http://localhost:8000',
      },
    });
  }

  const runId = client.getRunId();
  if (!runId) {
    fail({ error: 'SCP startRun did not return run_id' });
  }

  const display = formatCliForBuffer(cli!);
  appendToBuffer(terminalId, display, projectRoot);

  const now = new Date().toISOString();
  const session: SessionMetadata = {
    terminal_id: terminalId,
    created_at: now,
    cwd: process.cwd(),
    command: `clrun scp ${baseUrl}`,
    shell: '',
    status: 'running',
    pid: 0,
    worker_pid: 0,
    queue_length: 0,
    last_exit_code: null,
    last_activity_at: now,
    scp_run_id: runId,
    scp_base_url: baseUrl.replace(/\/$/, ''),
  };
  writeSession(session, projectRoot);

  logEvent('session.created', projectRoot, terminalId, {
    command: session.command,
    scp_run_id: runId,
    scp_base_url: session.scp_base_url,
  });

  success({
    terminal_id: terminalId,
    status: 'running',
    output: display,
    hints: {
      send_input: `clrun ${terminalId} "<option number or action name>"`,
      view_output: `clrun tail ${terminalId} --lines 50`,
      kill: `clrun kill ${terminalId}`,
      note: 'CLRUN supports dynamic remote CLIs via SCP. Send option index (e.g. "1") or action name to transition.',
    },
  });
}

File: python/clrun/commands/scp.py
"""The `clrun scp` command — connect to an SCP server and start a dynamic remote CLI session."""

from __future__ import annotations

import os

from clrun.utils.paths import resolve_project_root, ensure_clrun_dirs
from clrun.utils.output import success, fail
from clrun.runtime.lock_manager import acquire_lock
from clrun.pty.pty_manager import generate_terminal_id, write_session, create_session_metadata
from clrun.queue.queue_engine import init_queue
from clrun.buffer.buffer_manager import init_buffer, append_to_buffer
from clrun.ledger.ledger import log_event
from clrun.scp.client import SCPClient, format_cli_for_buffer


def scp_connect_command(base_url: str) -> None:
    """Connect to an SCP server and start a run. CLRUN supports dynamic remote CLIs via SCP."""
    project_root = resolve_project_root()
    ensure_clrun_dirs(project_root)
    acquire_lock(project_root)

    terminal_id = generate_terminal_id()
    init_queue(terminal_id, project_root)
    init_buffer(terminal_id, project_root)

    client = SCPClient(base_url)
    try:
        client.start_run()
        cli = client.get_cli()
    except Exception as e:
        fail({
            "error": f"SCP connect failed: {str(e)}",
            "hints": {
                "check_url": "Ensure the SCP server is running and reachable.",
                "example": "clrun scp http://localhost:8000",
            },
        })
        return

    run_id = client.get_run_id()
    if not run_id:
        fail({"error": "SCP start_run did not return run_id"})
        return

    display = format_cli_for_buffer(cli)
    append_to_buffer(terminal_id, display, project_root)

    base_normalized = base_url.rstrip("/")
    cwd = os.getcwd()
    session = create_session_metadata(
        terminal_id,
        f"clrun scp {base_url}",
        cwd,
        pid=0,
        worker_pid=0,
    )
    session.scp_run_id = run_id
    session.scp_base_url = base_normalized
    write_session(session, project_root)

    log_event("session.created", project_root, terminal_id, {
        "command": session.command,
        "scp_run_id": run_id,
        "scp_base_url": session.scp_base_url,
    })

    success({
        "terminal_id": terminal_id,
        "status": "running",
        "output": display,
        "hints": {
            "send_input": f"clrun {terminal_id} \"<option number or action name>\"",
            "view_output": f"clrun tail {terminal_id} --lines 50",
            "kill": f"clrun kill {terminal_id}",
            "note": "CLRUN supports dynamic remote CLIs via SCP. Send option index (e.g. \"1\") or action name to transition.",
        },
    })
